File: Apps/iOS/Domain/Entities/PositionOnHarmonica.swift
struct PositionOnHarmonica: Equatable {
    let fractionFromLeftEdge: Double
    let fractionAboveCentreLine: Double

    static func topmost(of positions: [PositionOnHarmonica]) -> PositionOnHarmonica? {
        positions.max { $0.fractionAboveCentreLine < $1.fractionAboveCentreLine }
    }
}

File: Web/src/domain/playing/positionOnHarmonica.ts
export interface PositionOnHarmonica {
    readonly fractionFromLeftEdge: number
    readonly fractionAboveCentreLine: number
}

export function topmostOf(positions: readonly PositionOnHarmonica[]): PositionOnHarmonica | null {
    let topmost: PositionOnHarmonica | null = null
    for (const position of positions) {
        if (topmost === null || position.fractionAboveCentreLine > topmost.fractionAboveCentreLine) {
            topmost = position
        }
    }
    return topmost
}

export function isOnTheHarmonica(position: PositionOnHarmonica): boolean {
    return position.fractionFromLeftEdge >= 0 && position.fractionFromLeftEdge <= 1
}
